File: ein.py/src/ein/cli/_factdump.py
"""Shared fact-dump helpers for the ``ein solve --print-final-*`` options.

Render a solution / final-state KB slice as canonical s-expressions — one dump
per solution branch, or the unsat-core facts when there is no model.
"""
from __future__ import annotations

from typing import Any

from ein.kb.store import KnowledgeBase


def fact_sexpr(arg: Any) -> str:
    """Render a Fact arg as an s-expression, recursing into nested
    Fact args (e.g. for ``(not (co-located Blue Green))``)."""
    from ein.kb.entities import Fact

    if isinstance(arg, Fact):
        inner = " ".join(fact_sexpr(a) for a in arg.args)
        return f"({arg.relation_name} {inner})" if inner else f"({arg.relation_name})"
    return str(arg)


def hypothesis_target_relations(kb: KnowledgeBase) -> set[str]:
    """Relations a query ``:hrules`` clause targets — the hypothesis
    commitments (e.g. zebra2's five ``*-loc``). Generic: the atoms named in
    the ``:hrules`` activators that are *declared relations*, so type/object
    atoms (``House``, ``Color``) drop out position-independently. Empty when
    the puzzle has no query/hrules."""
    if kb is None or kb.query is None:
        return set()
    from ein.ir import Atom, SForm

    def _atoms(node: Any, out: set[str]) -> set[str]:
        if isinstance(node, Atom):
            out.add(node.name)
        elif isinstance(node, SForm):
            if isinstance(node.head, Atom):
                out.add(node.head.name)
            for a in node.args:
                _atoms(a, out)
        return out

    names: set[str] = set()
    for kp in kb.query.kw_pairs:
        if kp.key.name == "hrules":
            _atoms(kp.value, names)
    return names & set(kb.relations)


def print_final_state(
    kb: KnowledgeBase, *, mode: str = "all",
    targets: set[str] | None = None,
) -> None:
    """Dump a slice of a solution kb's facts in canonical order.

    Three modes (the three ``--print-final-*`` flags):

    - ``all`` — the full REASONING layer: the propositional residue of the
      solve.
    - ``positive`` — ``all`` with the ``(not …)`` facts dropped too: the
      positive residue.
    - ``hfacts`` — only the positive facts whose relation is a query
      ``:hrules`` target (``targets``), across *every* layer so the given
      conditions count too — the hypothesis commitments (e.g. zebra2's 25
      ``*-loc``). Negatives have relation_name ``not`` ∉ targets, so they drop
      out for free.
    """
    from ein.kb.entities import Layer

    if mode == "hfacts":
        targets = targets or set()
        facts = [f for f in kb.facts if f.relation_name in targets]
        label = f"positive hypothesis-relation facts; :hrules {sorted(targets)}"
    else:
        facts = [
            f for f in kb.facts
            if f.layer == Layer.REASONING
            and not (mode == "positive" and f.relation_name == "not")
        ]
        label = (
            "REASONING layer, (not …) omitted" if mode == "positive"
            else "REASONING layer"
        )
    facts.sort(key=lambda f: (
        f.relation_name,
        tuple(fact_sexpr(a) for a in f.args),
    ))
    print(f"final-state facts ({label}; {len(facts)} facts):")
    for f in facts:
        args_str = " ".join(fact_sexpr(a) for a in f.args)
        print(f"  ({f.relation_name} {args_str})")

File: ein.rs/crates/ein-cli/src/factdump.rs
//! `ein solve --print-final-*` — a solution KB as canonical s-expressions.
//!
//! The Rust half of `ein/cli/_factdump.py`. `fact_sexpr` is *not* re-written
//! here: it is the event protocol's fact renderer, ported at
//! [P1a.0](../../../../docs/history/m1a_rust/README.md#p1a0--conformance-harness-and-shared-assets) as
//! `ein_infer::events::sexpr`, and the stage plan says reuse rather than
//! duplicate.

use ein_core::{Kb, Symbol, Terms};
use ein_infer::events::sexpr;
use ein_ir::{Ast, Node, NodeId};

/// Relations a query `:hrules` clause targets — the hypothesis commitments.
///
/// Generic: the atoms named in the `:hrules` activators that are *declared
/// relations*, so type/object atoms drop out position-independently. Every
/// `:hrules` pair contributes, not just the first — ein.py accumulates.
pub fn hypothesis_target_relations(ast: &Ast, terms: &Terms, kb: &Kb) -> Vec<Symbol> {
    let Some(query) = kb.program().query.as_ref() else {
        return Vec::new();
    };
    let mut names: Vec<String> = Vec::new();
    for &pair in query.kw_pairs.iter() {
        let Node::KwPair { key, value } = ast.node(NodeId(pair.0)) else {
            continue;
        };
        let Node::Keyword(name) = ast.node(key) else {
            continue;
        };
        if ast.sym(name) == "hrules" {
            collect_atoms(ast, value, &mut names);
        }
    }
    // `names & set(kb.relations)` — the intersection, then in *registry*
    // order, because the caller only ever sorts it or tests membership.
    let mut out: Vec<Symbol> = Vec::new();
    for (sym, _) in kb.program().relations.iter() {
        if names.iter().any(|n| n == terms.sym(sym)) {
            out.push(sym);
        }
    }
    out
}

/// `_atoms` — every atom name in a node: an `Atom`'s own, an `SForm`'s head
/// when it is an atom, and recursively its arguments. Nothing else.
fn collect_atoms(ast: &Ast, node: NodeId, out: &mut Vec<String>) {
    match ast.node(node) {
        Node::Atom(name) => out.push(ast.sym(name).to_string()),
        Node::SForm { head, args } => {
            if let Some(name) = ast.atom_name(head) {
                out.push(name.to_string());
            }
            for &a in ast.args(args) {
                collect_atoms(ast, a, out);
            }
        }
        _ => {}
    }
}

/// The three `--print-final-*` modes.
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Mode {
    /// `-p` — everything the engine derived: the propositional residue.
    All,
    /// `-P` — `All` with the `(not …)` facts dropped: the positive residue.
    Positive,
    /// `-f` — only the query `:hrules` target facts, *whatever their origin*,
    /// so the given conditions count too.
    Hfacts,
}

/// Dump a slice of a solution kb's facts in canonical order.
pub fn print_final_state(terms: &Terms, kb: &Kb, mode: Mode, targets: &[Symbol]) {
    let (mut facts, label): (Vec<_>, String) = match mode {
        Mode::Hfacts => {
            let picked: Vec<_> = kb
                .facts()
                .filter(|&f| targets.contains(&terms.facts.get(f).0))
                .collect();
            let mut names: Vec<&str> = targets.iter().map(|&s| terms.sym(s)).collect();
            names.sort_unstable();
            let shown: Vec<String> = names
                .iter()
                .map(|n| ein_core::pyrepr::repr_str(n))
                .collect();
            (
                picked,
                format!(
                    "positive hypothesis-relation facts; :hrules [{}]",
                    shown.join(", ")
                ),
            )
        }
        _ => {
            let not = terms.syms.get("not");
            let picked: Vec<_> = kb
                .facts()
                .filter(|&f| is_derived(kb, terms, f))
                .filter(|&f| !(mode == Mode::Positive && Some(terms.facts.get(f).0) == not))
                .collect();
            (
                picked,
                if mode == Mode::Positive {
                    "derived facts, (not …) omitted".to_string()
                } else {
                    "derived facts".to_string()
                },
            )
        }
    };

    // `key=(relation_name, tuple(fact_sexpr(a) for a in args))` — the args
    // rendered, not the raw ids, so the order is the one a reader sees.
    let mut keyed: Vec<(&str, Vec<String>, ein_core::FactId)> = facts
        .drain(..)
        .map(|f| {
            let (rel, args) = terms.facts.get(f);
            let rendered = args
                .iter()
                .map(|a| ein_infer::events::sexpr_value(terms, *a))
                .collect();
            (terms.sym(rel), rendered, f)
        })
        .collect();
    keyed.sort_by(|a, b| (a.0, &a.1).cmp(&(b.0, &b.1)));

    println!("final-state facts ({label}; {} facts):", keyed.len());
    for (rel, args, _) in &keyed {
        // Note the unconditional separator: a nullary fact prints `(rel )`
        // here, where `fact_sexpr` would spell it `(rel)`. ein.py builds this
        // line with its own join, and the difference is observable.
        println!("  ({rel} {})", args.join(" "));
    }
}

/// `Fact.is_derived` — produced by the engine, i.e. it has provenance and that
/// provenance is not a `:source` ingestion. A fact with none is neither given
/// nor derived, so it drops out of every `--print-final-*` mode but `hfacts`.
fn is_derived(kb: &Kb, terms: &Terms, fact: ein_core::FactId) -> bool {
    match kb.primary(fact) {
        Some(p) => terms.provs.get(p).kind != ein_core::ProvKind::Source,
        None => false,
    }
}

/// The unsat-core dump `--print-final-*` prints when there is no model.
pub fn print_unsat_core(terms: &Terms, core: &[ein_core::FactId]) {
    let mut keyed: Vec<(&str, Vec<String>, String)> = core
        .iter()
        .map(|&f| {
            let (rel, args) = terms.facts.get(f);
            let rendered: Vec<String> = args
                .iter()
                .map(|a| ein_infer::events::sexpr_value(terms, *a))
                .collect();
            (terms.sym(rel), rendered, sexpr(terms, f))
        })
        .collect();
    keyed.sort_by(|a, b| (a.0, &a.1).cmp(&(b.0, &b.1)));
    println!();
    println!("unsat-core facts ({} facts):", keyed.len());
    for (_, _, text) in &keyed {
        println!("  {text}");
    }
}
